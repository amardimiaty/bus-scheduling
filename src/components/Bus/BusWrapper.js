// packages
import styled from 'styled-components'

const BusWrapper = styled.div`
  position: relative;
  border-top: 1px solid #BDBDBD;
  height: 50px;
  cursor: pointer;

  .Bus__info {
    width: 140px;
    border-right: 1px solid #BDBDBD;
  }
  
  .Bus__number {
    display: inline-block;
    line-height: 50px;
    padding: 0 20px;
    font-weight: 600;
  }

  .Bus__time {
    font-size: 10px;
    font-weight: 400;
    color: #9E9E9E;
  }

  &:nth-child(odd) {
    background-color:#F5F5F5;
  }
`
export default BusWrapper