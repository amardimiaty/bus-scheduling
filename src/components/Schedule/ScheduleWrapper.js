// Packages
import styled from 'styled-components'


const ScheduleWrapper = styled.div`
  font-family: 'Lato';
  padding-top: 48px;
  width: 725px;
  margin: 0 auto;
  border-bottom: 1px solid #BDBDBD;
  .Schedule__operating-hrs {
    position: relative;
    left: 140px;
    font-size: 12px;
    margin-bottom: 12px;
    span {
      display: inline-block;
      width: 25px;
    }
    span:not(:first-child) {
      // Need to figure out better way to derive this #
      margin-left: 33px;
    }
  }
`

export default ScheduleWrapper