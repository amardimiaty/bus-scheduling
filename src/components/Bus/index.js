// Packages
import React, { useContext } from 'react'

// Utils
import formatTime from '../../utils/formatTime'

// Context
import { ScheduleContext } from '../Schedule'

// Styled Components
import BusWrapper from './BusWrapper'

// Components
import Trip from '../Trip'

const Bus = ({ trips, number }) => {
  const context = useContext(ScheduleContext)

  const getTripToAdd = () => {
    return context.schedule.filter(trip => trip.id === context.selectedTrip.id)[0]
  }

  /**
   * Takes in a trip to add to bus, then checks for 
   * conflicts by seeing if the trips start or end time falls
   * within 
   * 
   * @param {Object} tripToAdd trip object with startTime, endTime.
   * 
   * @returns {Boolean}
   */
  const checkForConflict = (tripToAdd) => {
    const { startTime, endTime } = tripToAdd
    for (let i = 0; i < trips.length; i++) {
      if (startTime >= trips[i].startTime && startTime <= trips[i].endTime) {
        return true
      }
      if (endTime >= trips[i].startTime && endTime <= trips[i].endTime) {
        return true
      }
    }
    return false
  } 

  /**
   * Gets the selected trip checks for scheduling
   * conflict with bus we want to add it to.
   * 
   * @returns {(Boolean | Object)} 
   * false if there is a conflict, if not an updated bus obj
   */
  const updateBusWithNewTrip = () => {
    const tripToAdd = getTripToAdd()
    const hasConflict = checkForConflict(tripToAdd)
    if(hasConflict) {
      return false
    }
    // Sort trips by start time
    const newTrips = [...trips, tripToAdd].sort((a,b) => a.startTime - b.startTime)
    return { number, trips: newTrips }
  }

  /**
   * Removes the selected trip from the current bus and
   * then adds trip to new bus. Last remove buses with
   * no trips and reset the bus numbers.
   * 
   * @returns {Array} The updated buses array.
   */
  const getUpdatedBuses = () => {
    const updatedBus = updateBusWithNewTrip()

    if(updatedBus) {
      return context.buses.map(bus => {
        // remove from current bus
        if (context.selectedTrip.currentBus === bus.number) {
          const trips = bus.trips.filter(trip => trip.id !== context.selectedTrip.id)
          return bus = { number, trips: trips }
        }
        // update bus
        if (bus.number === number) {
          return bus = updatedBus
        }

        return bus

      }).filter(bus => bus.trips.length !== 0)
        .map((bus, idx) => {
          return {
            number: idx + 1,
            trips: bus.trips,
          }
        })
    } else {
      // Since there is a conflict remove the "New Bus" and alert
      alert('There is a scheduling conflict cannot add trip to bus.')
      return context.buses.slice(0, context.buses.length - 1)
    }
  }
  
  const handleBusClick = ({ target }) => {
    if(target.classList.contains('Bus')) {
      if(context.selectedTrip.id) {
        const sameBus = trips.filter(t => t.id === context.selectedTrip.id )
        if(!sameBus.length) {
          const updatedBuses = getUpdatedBuses()
          context.updateBuses(updatedBuses)
          context.updateSelected({})
        }
      }
    }
  }

  return (
      <BusWrapper className="Bus" onClick={(e) => handleBusClick(e)}>
        <div className="Bus__info">
          <span className="Bus__number">
            {typeof number === "number" ? `Bus ${number}` : 'New Bus'}
          </span>
          <span className="Bus__time">
            {
              trips.length ? (
                `${formatTime(trips[0].startTime)} 
                - 
                ${formatTime(trips[trips.length - 1].endTime)}`
              ) : null
            }
          </span>
        </div>
        {
          trips.map(trip => (
            <Trip key={`trip-${trip.id}`} trip={trip} busNumber={number} />
          ))
        }
      </BusWrapper>
  )
}

export default Bus