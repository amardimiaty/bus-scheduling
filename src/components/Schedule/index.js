// Packages
import React, { useState, useEffect, createContext } from 'react'

// Data
import BusSchedule from '../../data/schedule.json'
import OperatingHrs from '../../data/operatingHours.json'

// Styled Components
import ScheduleWrapper from './ScheduleWrapper'

// Components
import Bus from '../Bus'

const ScheduleContext = createContext()
const ScheduleProvider = ScheduleContext.Provider

const Schedule = () => {
  const schedule = BusSchedule
  const operatingHours = OperatingHrs
  const [buses, updateBuses] = useState([])
  const [selectedTrip, updateSelected] = useState({})
  const [isNewBusShowing, updateIsNewBusShowing] = useState(false)

  /* eslint-disable */
  useEffect(() => {
    // Initial seeding.
    let busesToAdd = []
    schedule.map((trip, idx) => busesToAdd.push({number: idx + 1,  trips: [trip] }))
    updateBuses(busesToAdd)
  }, [])

  useEffect(() => {
    // If the selected trip changes and it is not empty show new bus.
    if(selectedTrip.id) {
      if(!isNewBusShowing) {
        const busesWithNew = [...buses, { number: 'New Bus', trips: [] }]
        updateBuses(busesWithNew)
        updateIsNewBusShowing(true)
      }
    } else {
      updateIsNewBusShowing(false)
    }
  }, [selectedTrip])
/* eslint-enable */

  return (
    <ScheduleProvider value={{
      schedule,
      selectedTrip, 
      updateSelected,
      buses,
      updateBuses,
    }}>
      <ScheduleWrapper>
        <div className="Schedule__operating-hrs">
          {
            operatingHours.map(h => (<span key={h}>{h}</span>))
          }
        </div>
        {
          buses.map((bus, idx) => (
            <Bus key={`bus-${idx + 1}`} trips={bus.trips} number={bus.number}/>
          ))
        }
      </ScheduleWrapper>
    </ScheduleProvider>
  )
}

export default Schedule
export { ScheduleContext, ScheduleProvider }