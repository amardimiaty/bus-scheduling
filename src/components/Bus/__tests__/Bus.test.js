// Packages
import React from 'react'
import wait from 'waait'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'

// Context
import { ScheduleProvider } from '../../Schedule'

// Components
import Bus from '../index'

const fakeTrips = [
  { 'id': 1, 'startTime': 30, 'endTime': 150 }
]

const fakeContext = {
  schedule : [
    { 'id': 1, 'startTime': 30, 'endTime': 150 },
  ],
  selectedTrip: {},
  updateSelected: () => {},
  buses: [
    { number: 1, trips: [{ 'id': 1, 'startTime': 30, 'endTime': 150 }] }
  ],
  updateBuses: () => {},
}

describe('<Bus />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <ScheduleProvider value={fakeContext}>
        <Bus number={1} trips={fakeTrips} />
      </ScheduleProvider>
    )
    await wait()
    wrapper.update()
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it('renders the correct bus number', async () => {
    const wrapper = mount(
      <ScheduleProvider value={fakeContext}>
        <Bus number={1} trips={fakeTrips} />
      </ScheduleProvider>
    )
    await wait()
    wrapper.update()
    const busNumber = wrapper.find('.Bus__number')
    expect(busNumber.text()).toContain('Bus 1');
  });
});