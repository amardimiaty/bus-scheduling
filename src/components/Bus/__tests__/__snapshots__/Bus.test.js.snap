// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bus /> renders and matches snapshot 1`] = `
<Bus
  number={1}
  trips={
    Array [
      Object {
        "endTime": 150,
        "id": 1,
        "startTime": 30,
      },
    ]
  }
>
  <styled.div
    className="Bus"
    onClick={[Function]}
  >
    <StyledComponent
      className="Bus"
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bwzfXH",
            "isStatic": false,
            "lastClassName": "iqFWvV",
            "rules": Array [
              "
  position: relative;
  border-top: 1px solid #BDBDBD;
  height: 50px;
  cursor: pointer;

  .Bus__info {
    width: 140px;
    border-right: 1px solid #BDBDBD;
  }
  
  .Bus__number {
    display: inline-block;
    line-height: 50px;
    padding: 0 20px;
    font-weight: 600;
  }

  .Bus__time {
    font-size: 10px;
    font-weight: 400;
    color: #9E9E9E;
  }

  &:nth-child(odd) {
    background-color:#F5F5F5;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bwzfXH",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={[Function]}
    >
      <div
        className="sc-bwzfXH iqFWvV Bus"
        onClick={[Function]}
      >
        <div
          className="Bus__info"
        >
          <span
            className="Bus__number"
          >
            Bus 1
          </span>
          <span
            className="Bus__time"
          >
            0:30 
                - 
                2:30
          </span>
        </div>
        <Trip
          busNumber={1}
          key="trip-1"
          trip={
            Object {
              "endTime": 150,
              "id": 1,
              "startTime": 30,
            }
          }
        >
          <styled.div
            endTime={150}
            isSelected={false}
            onClick={[Function]}
            startTime={30}
          >
            <StyledComponent
              endTime={150}
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-htpNat",
                    "isStatic": false,
                    "lastClassName": "jHRzPn",
                    "rules": Array [
                      "
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4px;
  height: 30px;
  border: 1px solid #9E9E9E;
  width: ",
                      [Function],
                      "px;
  left: ",
                      [Function],
                      "px;
  background-color: ",
                      [Function],
                      ";
  cursor: pointer;
",
                    ],
                  },
                  "displayName": "styled.div",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-htpNat",
                  "target": "div",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
              isSelected={false}
              onClick={[Function]}
              startTime={30}
            >
              <div
                className="sc-htpNat jHRzPn"
                onClick={[Function]}
              >
                <span
                  className="Trip__number"
                >
                  1
                </span>
              </div>
            </StyledComponent>
          </styled.div>
        </Trip>
      </div>
    </StyledComponent>
  </styled.div>
</Bus>
`;
